e2e/e2eutil/e2eutil.go

```diff
@@ -6,174 +6,108 @@ package e2eutil
-		LabelSelector: "app.kubernetes.io/managed-by=e2eutil",
+var Purge = func(ctx context.Context, cfg *envconf.Config) (context.Context, error) {
+	client, err := cfg.NewClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create client during setup/purge: %s", err)
+	}
+	var list corev1.NamespaceList
+	err = client.Resources().List(ctx, &list, func(opts *metav1.ListOptions) {
+		opts.LabelSelector = "app.kubernetes.io/managed-by=e2eutil"
 	})
+
 	if err != nil {
 		fmt.Printf("error while listing namespaces to purge past test runs: %s\n", err)
-		return
+		return ctx, err
 	}
 	for _, item := range list.Items {
 		fmt.Printf("cleaning up namespace from previous run %q\n", item.Name)
-		err = s.Clientset.CoreV1().Namespaces().Delete(context.Background(), item.Name, metav1.DeleteOptions{})
+		err = client.Resources().Delete(ctx, &item)
 		if err != nil {
 			fmt.Printf("error while cleaning up namespace %q: %s\n", item.Name, err)
+			return ctx, err
 		}
 	}
-}
 
-type Case struct {
-	*Suite
-	t  *testing.T
-	ns string
-}
+	// cleanup Prometheus cluster-level resources
+	promClusterRoleBinding := &rbacv1.ClusterRoleBinding{ObjectMeta: metav1.ObjectMeta{
+		Name: fixtures.PromServer,
+	}}
+	if err := client.Resources().Delete(ctx, promClusterRoleBinding); err != nil {
+		return ctx, err
+	}
 
-func (c *Case) Retry(fn func() error) {
-	for {
-		err := fn()
-		if err == nil {
-			return
-		}
-		log.Printf("error: %s", err)
-		time.Sleep(util.Jitter(time.Millisecond*200, 0.5))
+	// cleanup Prometheus cluster-level resources
+	promClusterRole := &rbacv1.ClusterRole{ObjectMeta: metav1.ObjectMeta{
+		Name: fixtures.PromServer,
+	}}
+	if err := client.Resources().Delete(ctx, promClusterRole); err != nil {
+		return ctx, err
 	}
+
+	return ctx, nil
 }
 
-func (c *Case) Hostname(domain string) string {
-	c.ensureNS()
-	return strings.ToLower(c.ns) + "." + domain
+func GetHostname(ns, domain string) string {
+	return strings.ToLower(ns) + "." + domain
 }
 
-// WithResources creates Kubernetes resources for the test case and waits for them to become ready.
-func (c *Case) WithResources(resources ...client.Object) {
-	c.ensureNS()
-
-	var wg sync.WaitGroup
-	for _, res := range resources {
-		wg.Add(1)
-		go func(res client.Object) {
-			defer wg.Done()
-
-			res.SetNamespace(c.ns)
-			c.Retry(func() error {
-				err := c.Client.Create(context.Background(), res)
-				if err != nil && k8serrors.IsAlreadyExists(err) {
-					err = c.Client.Update(context.Background(), res)
-				}
-
-				return err
-			})
-
-			switch obj := res.(type) {
-			case *appsv1.Deployment:
-				c.watchDeployment(obj)
-			}
-		}(res)
-	}
-	wg.Wait()
+// CreateNSForTest creates a random namespace with the runID as a prefix. It is stored in the context
+// so that the deleteNSForTest routine can look it up and delete it.
+func CreateNSForTest(ctx context.Context, cfg *envconf.Config, t *testing.T) (context.Context, error) {
+	prefix := "e2e-" + strings.ToLower(t.Name()) + "-"
```

Comment:
we'd rather take advantage of guaranteed name uniqueness. We can do that with the `GenerateName` field on the Namespace object instead of the `Name` field. 

This function will become

1. create ns with `GenerateName` field
2. if create was a success we use that name

This is guaranteed versus the randomString approach. Also a lot cleaner

Reply:
good call, added it in